### ISSUE TITLE
Analyzing and Transforming Data - Empty Box notation ambigous 

### ISSUE THREAD

**Issue by backface** [jmoenig/Snap--Build-Your-Own-Blocks#2189]
I tried to recreate the analyze block from the "Tables in Snap!" Manual update (https://snap.berkeley.edu/snapsource/TablesInSnap.pdf) and it did not work.

On further examination, It seems to me the empty box notation with nested Higher Order List Operations is unclear and ambiguous, and I could make it work by being more explicit with input names - as demonstrated here:
https://snap.berkeley.edu/snapsource/snap.html#present:Username=mash&ProjectName=analyze&editMode

But I guess, empty box notation should be avoided with nesting functions (or in general) and examples adapted.

**Comment by DyslexicAwe** [comment]
Can you explain the custom block   
![unknown block](https://user-images.githubusercontent.com/16387241/44576109-08d26180-a78e-11e8-8154-5fa6b4d5da7c.png)
which you are using - I tried to edit it in order to see what's inside, but I can't.

**Comment by DyslexicAwe** [comment]
@towerofnix Thanks. Is this something new or it's just that I have not noticed these gray ones among all the other green ones in the Operators category? I was not aware of the possibility of Boolean rings with inputs, at all, (and I don't remember seeing it mentioned in the Manual or its updates).  
![photoeditor_20180826_141731466](https://user-images.githubusercontent.com/16387241/44628217-d5cdd080-a93b-11e8-8106-5a4eaad7b5d3.jpg)

**Comment by towerofnix** [comment]
@DyslexicAwe You just haven't noticed - those have been around for a very long time. I don't think the manual explicitly discusses them, but they're basically like "lambdas" in Python (e.g. `lambda x: x + x * x`) and "inline functions" in JavaScript (e.g. `const array = [ function() {...}, function() {...} ]`).

**Comment by brianharvey** [comment]
Gray rings are introduced on page 31 of the manual, but the extended discussion is in Chapter VI, Procedures as Data, starting on page 43.  The reason neither of you noticed this is that it's actually pretty uncommon to drag a ring explicitly from the Operators palette.  Most of the time, the rings are already in the input slots of the blocks where they're needed.  If you need to add or remove a ring manually, you use "ringify" or "unringify" as appropriate.  The only explicit mention of the palette rings are on page 44.  (In my defense I point out that looking up "ring" in the index will point you there.)

**Comment by backface** [comment]
Honestly, I do not agree it's a duplicate. My initial point is that the "analyze" example demonstrated in the "Tables in Snap!" Manual update (page 15) simply does not work. Only when naming the inputs I can make it work!

**Comment by jguille2** [comment]
Yes, I think this is another issue.

For me the problem is only in the _Tables in Snap!_ manual, not in the feature.

White slots are related to their 'mapping' function inputs (if you don't set a specific name for those). But if you have different 'mapping' functions and the inner function use the arguments of the outer function, you have to name that arguments.

So, I think the _analyze_ example on _Tables in Snap!_ must be:
![analyze](https://user-images.githubusercontent.com/2926315/44951587-398a5900-ae68-11e8-9a44-5c30ec1e76b8.png)

The inner "mapping" function (keep) input can be named, but it is not necessary because we don't use it elsewhere.

**Comment by DyslexicAwe** [comment]
Aha, now I understand. Thanks, @jguille2 

I will rephrase what I think you were saying: "But if you have **nested** (_I think this is what you meant by 'different'_) 'mapping' functions and the inner function uses the arguments _from_ the outer function, you have to name that arguments."

If I understand correctly, there are actually two levels of nestedness in this particular case as I have visualized, picturing the "data flow" to understand what you said. So the cyan-colored arrow is the most outer level, then darker blue arrows are the first level of nestedness and the red arrow is the second level of nestedness. Is this correct?
![image 1](https://user-images.githubusercontent.com/16387241/44954689-475ecf00-aea6-11e8-90a4-766f4bbcd6d7.png)

**Comment by backface** [comment]
Agreed: Definitely, It's the analyze example in the manual that should be fixed.

To rephrase the main question (I was always talking of nested functions):
If you have nested mapping functions, what is the defined order in which they fill the empty slots?

And I was wondering - as the example is there as it is: Is it simply wrong or did it work at some time, which means the order did change?

**Comment by brianharvey** [comment]
Since this example is just in Jens's supplementary document, which is really obsolete now that the manual has caught up, I still think there's nothing we have to fix here.  But, to answer your question, nothing has changed about the interpretation of empty slots, so the example was wrong all along, I guess.
